src/compartment.c

```diff
@@ -0,0 +1,526 @@
+                                    to_map->fd, curr_seg->offset);
+        }
+        else
+        {
+            assert(curr_seg->mem_sz > curr_seg->file_sz);
+            map_result = mmap((void*) curr_seg->mem_bot,
+                                    curr_seg->mem_sz,
+                                    curr_seg->prot_flags,
+                                    MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS,
+                                    -1, 0);
+            assert(map_result !=  MAP_FAILED);
+            int pread_res = pread(to_map->fd, (void*) curr_seg->mem_bot,
+                                  curr_seg->file_sz, curr_seg->offset);
+            assert(pread_res != -1);
+        }
+    }
+
+    printf("Currently mapping compartment scratch memory + stack\n");
+    map_result = mmap((void*) to_map->scratch_mem_base,
+                      to_map->scratch_mem_size,
+                      PROT_READ | PROT_WRITE,
+                      MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS,
+                      -1, 0);
+    assert(map_result != MAP_FAILED);
+    map_result = mmap((void*) to_map->scratch_mem_stack_top - to_map->scratch_mem_stack_size,
+                      to_map->scratch_mem_stack_size,
+                      PROT_READ | PROT_WRITE,
+                      MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS | MAP_STACK,
+                      -1, 0);
+    to_map->stack_pointer = to_map->scratch_mem_stack_top;
+    assert(map_result != MAP_FAILED);
+
+    for (size_t i = 0; i < to_map->relas_cnt; ++i)
+    {
+        uintptr_t rela_addr = to_map->relas[i] + to_map->base;
+        uintptr_t old_plt_val = (uintptr_t) *((void**) rela_addr);
+        old_plt_val += to_map->base;
+        *((uintptr_t *) rela_addr) = old_plt_val;
+    }
+
+    for (size_t i = 0; i < to_map->curr_intercept_count; ++i)
+    {
+        struct intercept_patch to_patch = to_map->patches[i];
+        *to_patch.patch_addr = to_patch.instr;
+    }
+
+    to_map->mapped = true;
+}
+
+void
+comp_map_full(struct Compartment* to_map)
+{
+    assert(!(to_map->mapped || to_map->mapped_full));
+    void* map_result = mmap((void*) to_map->base, to_map->size,
+        PROT_WRITE | PROT_READ | PROT_EXEC, // TODO
+        MAP_PRIVATE | MAP_FIXED,
+        to_map->fd, 0);
+    assert(map_result != MAP_FAILED);
+    to_map->mapped_full = true;
+}
+
+void ddc_set(void *__capability cap) {
+    assert(cap != NULL);
+    asm volatile("MSR DDC, %[cap]" : : [cap] "C"(cap) : "memory");
+}
+
+int64_t
+comp_exec(struct Compartment* to_exec)
+{
+    void* fn = (void*) to_exec->entry_point;
+    void* wrap_sp;
+
+    setup_stack(to_exec);
+
+    int64_t result;
+
+/*#if __has_feature(capabilities)*/
+    /*ddc_set(to_exec->ddc);*/
+/*#endif*/
+
+    // TODO handle register clobbering stuff (`syscall-restrict` example)
+    // https://github.com/capablevms/cheri_compartments/blob/master/code/signal_break.c#L46
+    asm("str lr, [sp, #-16]!\n\t"
+        "mov sp, %[comp_sp]\n\t"
```

Comment:
This needs to preserve the old `sp` somewhere so it can restore it. I can't see how the `wrap_sp` mechanism works. What I think this needs to do is to save the `sp` in some temporary register, then push it to the _new_ stack (`comp_sp`) so you can restore it when you return. Something like this:

```
str lr, [sp, #-16]!
mov x16, sp
mov sp, %[comp_sp]
str x16, [sp, #-16]!
...
ldr x16, [sp], #16
mov sp, x16
ldr lr, [sp], #16
```

Since you have to put two registers both times, you can probably put the old LR on the new stack too:

```
mov x16, sp
mov sp, %[comp_sp]
stp x16, lr, [sp, #-16]!
...
ldp x16, lr, [sp], #16
mov sp, x16
```

This approach only works under hybrid, because the boundary is defined by DDC and PCC. Under purecap, if you're storing capabilities, you're also leaking them across a compartment boundary. (In your earlier purecap example switcher, you had to put it in a special area belonging to the manager, if I remember correctly.)

Reply:
That's fine, of course! You presented it as a work-in-progress, after all.

It is worth considering this early in general terms, since with some switching mechanisms, it can be hard to arrange storing the bits you need to return. However, we already solved it one way for purecap, and I think it's a little easier for this hybrid arrangement.